### PR TITLE
lib: bsdlib: rename to nrf_modem

### PR DESCRIPTION
**Pull request by lemrey**
Rename `bsdlib` to `nrf_modem` (library) and ~~`libmodem`~~ `nrf_modem_lib` (glue).
reminder! The two commits must be squashed.

### CODE REVIEW

**Review by coderbyheart — APPROVED** [review]

File: doc/nrf/releases/release-notes-0.3.0.rst [hunk 1]
```diff
@@ -112,7 +112,7 @@ nRF9160
     This library helps creating an AT command socket and forwards requests and responses from and to the modem.
   * :ref:`lib_nrf_cloud`:
     This library implements features to connect and send data to nRF Cloud services.
-  * **bsdlib**:
+  * **nrf_modem**:
```
[the remark this answers is not in the document]
Comment (lemrey) [reply doc/nrf/releases/release-notes-0.3.0.rst:115]: @umapraseeda thinks so

[the remark this answers is not in the document]
Comment (umapraseeda) [reply doc/nrf/releases/release-notes-0.3.0.rst:115]: I meant to update the links (like nrfxlib:bsdlib to ref:`nrfxlib:nrf_modem`). May be retain this one and the other bsdlib headings in old release notes, @ru-fu ?

[the remark this answers is not in the document]
Comment (ru-fu) [reply doc/nrf/releases/release-notes-0.3.0.rst:115]: Yes, that's how we've done it for the SD Controller as well. Update the references in the release notes (because otherwise, the build will break), but leave the old text as it is.

**Comment by evenl** [comment]
I am a bit confused about the different naming conventions used for this library: 

Before with bsdlib:

Files: libbsd_nrf9160_xxaa.a, bsd_os.c, bsdlib.c, nrf91_sockets.c
Folder: bsdlib (both nrf and nrfxlib)
Kconfig nrfxlib: CONFIG_BSD_LIB_
Kconfig nrf:  CONFIG_BSD_LIBRARY_, CONFIG_NRF91_SOCKET

Today with modemlib:

File libmodem.a (nrfxlib), libmodem.c (nrf), nrf91_sockets.c (nrf), nrf_modem_os.c (nrf)
Folder: nrf_modem (nrfxlib), libmodem (nrf)
Kconfig nrfxlib: CONFIG_NRF_MODEM
Kconfig nrf: CONFIG_LIBMODEM, CONFIG_NRF91_SOCKETS_

I think it make more sense to call the folder in nrf for nrf_ modem (same as nrfxlib) instead of libmodem? I think it also make more sense to rename libmodem.c to nrf_modem.c, and the CONFIG_LIBMODEM kconfig parameters to CONFIG_NRF_MODEM_LIB_. I am fine with the .a file other that it should probably contain some device information like nrf9160 at least. 

A common convention is to name the library file (.a) libfoobar.a, but any references to the library is usually "foobar library" or "foobar lib". Of course both conventions can be found, but for me nrf_modem, nrf_modem_lib or nrf_modem_library make more sense for folders, files and kconfig paramters extern to the library it self, and align more with the naming convention used for bsdlib.

**Comment by lemrey** [comment]
@evenl The point was to create a distinction between the glue and the library. I think there needs to be one.
Here we have `libmodem.h`, `libmodem_init`, `libmodem.c` (under `/libmodem`) and it's clear that it's glue and it belongs to `sdk-nrf`.

Following your suggestion, if I renamed the glue to `nrf_modem` we'd end up with:
- `nrf_modem.h`, duplicate of main header in `nrfxlib/nrf_modem/include/nrf_modem.h` 
- `nrf_modem.c`, which would not the implementation of above header from `nrfxlib`
- `nrf_modem_init` in `sdk-nrf` `nrf_modem.c`, duplicate of `nrf_modem_init` in above header

So in conclusion it would be hard to make that distinction anymore (including the distinction between library-defined configurations and glue-defined configurations) and we'd have collisions.

**Comment by evenl** [comment]
Yes, I understand, but my point is that it is not clear to me that the libmodem prefixed stuff and the nrf_modem stuff belongs together. Its a bit confusing, for bsdlib it was bsd_ for the nrfxlib stuff, and bsdlib_ for the OS integration part. To me it would have been much more logical with for instance nrf_modem_ for the nrfxlib stuff and nrf_modem_lib_ for the OS integration part.

**Comment by rlubos** [comment]
I think that Even might have a point here, that `libmodem` might not be easily connected to `nrf_modem` for someone not familiarized with the nomenclature. `nrf_modem_lib` sounds to be better associated with the actual nrfxlib component, the name (and the prefix) is a bit long, but to be honest, nothing better comes to my mind.

**Comment by evenl** [comment]
Yes, I agree that it's a bit long, but same here, I can not come up with any better.

**Comment by lemrey** [comment]
I see, point taken. I can change to `nrf_modem_lib` though I preferred `libmodem` :)

**Comment by lemrey** [comment]
Renamed and rebased.

**Review by tejlmand — CHANGES_REQUESTED** [review]
We should really be careful with such Kconfig renaming that is proposed in this PR.

They might appear small to us, but they cause a lot of frustration for users of NCS when we change them.

File: lib/nrf_modem_lib/Kconfig [hunk 2]
```diff
@@ -3,31 +3,26 @@
...
-menu "BSD Library for nrf91"
-
-config BSD_LIBRARY
-	bool
-	prompt "Use BSD Socket library for IP/TLS/DTLS"
-	imply FPU
-	imply FPU_SHARING
-	select NET_SOCKETS_POSIX_NAMES
-	imply NET_SOCKETS_OFFLOAD
+menuconfig NRF_MODEM_LIB
```
[the remark this answers is not in the document]
Comment (evenl) [reply lib/nrf_modem_lib/Kconfig:6]: This change has be through quite a few peoples that has agreed to the change before this PR was created, but unfortunately it has not been through you before. This move has been planned for some time, and we are currently in a state where we are holding back libmodem production a bit while waiting for this and the nrfxlib PRs to be merged. The goal was to merge this to master early in the NCS 1.4.0 -> NCS 1.5.0 window to give the organization some time to catch up with the change.

If you are saying that we should not do it now we have kind of a problem. We have a fairly big queue of features and bug fixes that is waiting (in the new repo) for this test release to go thought, and going back to the old BSD library would create a fair amount of double work.

[the remark this answers is not in the document]
Comment (tejlmand) [reply lib/nrf_modem_lib/Kconfig:6]: that is very unfortunate, let me suggest that we keep the existing symbols and deprecate them instead.
I have made a suggestion here: 
https://github.com/lemrey/sdk-nrf/pull/9

Please add any other rename symbols in same manner.

If the deprecated symbol is selected, then we will automatically select the corresponding `NRF_MODEM` symbol.

This will look like:


it can then be decided later if / how we want to inform the user that he is using a deprecated symbol and which symbol he should be using instead.

[the remark this answers is not in the document]
Comment (lemrey) [reply lib/nrf_modem_lib/Kconfig:6]: Seems like a good idea, thanks.

[the remark this answers is not in the document]
Comment (lemrey) [reply lib/nrf_modem_lib/Kconfig:6]: @tejlmand I have updated the PR based on your suggestion, thank you.

**Review by tejlmand — APPROVED** [review]

File: lib/nrf_modem_lib/Kconfig (hunk 2 above)
[the remark this answers is not in the document]
Comment (evenl) [reply lib/nrf_modem_lib/Kconfig:6]: @tejlmand @lemrey Is this the best approach? Wouldn't it be better to then just keep bsdlib around instead, we could deprecate bsdlib at a later stage. We don't do any more work on bsdlib, and put a deprecated stamp on the kconfig parameters and and in the change log. And the two libraries would also have to be mutual exclusive.

Isn't this what we asked @carlescufi about earlier lemrey? Because @bjorn-tore-taraldsen was worried about a hard removal of bsdlib, and though we should keep bsdlib around for some time.

[the remark this answers is not in the document]
Comment (lemrey) [reply lib/nrf_modem_lib/Kconfig:6]: We just did a rename here, so we agreed that that did not warrant having to keep the old library in the repo.
It is okay to provide a set of compatible KConfig to smooth the transition of those, at least, since it's not much work anyway but we don't need to keep the old lib around.

[the remark this answers is not in the document]
Comment (evenl) [reply lib/nrf_modem_lib/Kconfig:6]: ok

**Review by evenl — APPROVED** [review]

File: doc/nrf/releases/release-notes-1.0.0.rst [hunk 3]
```diff
@@ -112,11 +112,11 @@ nRF9160
 * nRF Connect SDK now uses upstream CoAP implementation. The :ref:`mqtt_simple_sample` sample was rewritten to use the upstream library, and the downstream CoAP was removed.
 * The :ref:`http_application_update_sample` sample has been updated to use the :ref:`lib_fota_download` library.
 
-Modem library
+BSD library
 -------------
```
Comment (lemrey) [inline doc/nrf/releases/release-notes-1.0.0.rst:116]: I suppose these should stop the the same column as `BSD library`

Comment (umapraseeda) [reply doc/nrf/releases/release-notes-1.0.0.rst]: Fixed these..

**Review by umapraseeda — APPROVED** [review]

**Comment by lemrey** [comment]
Rebased again to try and get CI green.

**Review by eibe85 — APPROVED** [review]

File: west.yml [hunk 4]
```diff
@@ -102,7 +102,7 @@ manifest:
     - name: nrfxlib
       repo-path: sdk-nrfxlib
       path: nrfxlib
-      revision: ab60d3ca0a7befd2365d37fb19a7bd493d6b58a4
+      revision: pull/337/head
```
Comment (rlubos) [inline west.yml:105]: Please update SHA: `9908d7c173c5fc5695745614f7c556d786f838ce`

Comment (lemrey) [reply west.yml]: Done

**Review by rlubos — APPROVED** [review]
